drivers: wifi: siwx91x: Handle MFP configuration and SAE password validation

Nitin-Pandey-01 opened the pull request:
- Defined KConfig for enable/disable MFP because, by default MFP is enabled in opermode during wifi init.
- Modified conditions to store SAE password and PSK in different cases.

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -293,10 +325,21 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
 static int siwx91x_status(const struct device *dev, struct wifi_iface_status *status)
 {
 	struct siwx91x_dev *sidev = dev->data;
```
rebase the change

Nitin-Pandey-01 replied:
Done

Nitin-Pandey-01 commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -49,13 +49,32 @@ static unsigned int siwx91x_on_join(sl_wifi_event_t event,
 	return 0;
 }
 
+static enum wifi_mfp_options siwx91x_get_default_mfp(sl_wifi_security_t security)
```
Done, Passed interface as a default interface

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -67,6 +67,22 @@ static int siwx91x_security(enum wifi_security_type security)
+{
+	if (interface & SL_WIFI_CLIENT_INTERFACE) {
+		switch (security) {
+		case SL_WIFI_WPA3:
```
add case for SL_WIFI_WPA3_TRANSITION

Nitin-Pandey-01 replied:
Done

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -67,6 +67,22 @@ static int siwx91x_security(enum wifi_security_type security)
+		switch (security) {
+		case SL_WIFI_WPA3:
+			return WIFI_MFP_REQUIRED;
+		case SL_WIFI_WPA2:
```
Add for WPA_WPA2_MIXED for future

Nitin-Pandey-01 replied:
Done

Nitin-Pandey-01 commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -49,13 +49,32 @@ static unsigned int siwx91x_on_join(sl_wifi_event_t event,
+		default:
+			return WIFI_MFP_DISABLE;
+		}
+	} else {
```
Done

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -278,14 +278,22 @@ static int siwx91x_connect(const struct device *dev, struct wifi_connect_req_par
+	ret = SL_STATUS_FAIL;
+	if (params->sae_password_length >= WIFI_PSK_MIN_LEN &&
+	    params->sae_password_length <= WIFI_PSK_MAX_LEN) {
+		ret = sl_net_set_credential(SL_NET_DEFAULT_WIFI_CLIENT_CREDENTIAL_ID,
```
I expected `sl_net_set_credential()` return an error if wiseconnect was not able to handle the requested credentials. Does it make sense to sanitize the data?

Nitin-Pandey-01 replied:
For PSK, a validation check has been implemented in wifi_connect(). However, no such check exists for the SAE password. To prevent unnecessary rejections, I am sanitizing the data, what do you suggest?

jerome-pouiller replied:
Maybe we should rather patch `wifi_connect()`?

Nitin-Pandey-01 replied:
Sure, Updated the code as suggested

jerome-pouiller reviewed (commented):
Ideally, the commit log should explain what bug is fixed. Without this patch, what was the behavior when params->sae_password_length was too large/small?

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -257,6 +314,12 @@ static int siwx91x_connect(const struct device *dev, struct wifi_connect_req_par
+			LOG_WRN("Needed MFP %s but got MFP %s, hence setting to MFP %s",
+				wifi_mfp_txt(WIFI_MFP_REQUIRED), wifi_mfp_txt(params->mfp),
+				wifi_mfp_txt(WIFI_MFP_REQUIRED));
+			params->mfp = WIFI_MFP_REQUIRED;
```
Why this case is not managed by `siwx91x_configure_mfp_option()`?

Nitin-Pandey-01 replied:
This case is not handled by siwx91x_configure_mfp_option() because WPA2-SHA256 is internally mapped to WPA2. The key difference is that WPA2-SHA256 requires MFP to be set to 'required'.

- This  siwx91x_configure_mfp_option() function handles two cases with respect to WPA2,
    1. User configured WPA2-PSK and MFP required. -- Connection should happen in WPA2-SHA256
    2. User configured WPA2-PSK and MFP Optional -- Connection should happen in WPA2-SHA1
 -  The only case which need to handle separately is if User configured WPA-SHA256 and gave other MFP option instead of MFP required.
 
 Note:  The function uses wifi_config.security instead of params->security to avoid conflicts in mixed modes, where both WPA_WPA2 Mixed and WPA3 Transition modes are mapped to WPA_AUTO_PERSONAL. We will request SDK team for improvement.

jerome-pouiller replied:
I see. Then, I just suggest to move these line next the call to `siwx91x_configure_mfp_option()`. Thus, the handle of mfp is grouped in the same place adn this switch case will continue to only do a conversion.

Nitin-Pandey-01 replied:
Sure, Updated the code as suggested

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -487,6 +555,7 @@ static int siwx91x_status(const struct device *dev, struct wifi_iface_status *st
 {
 	sl_si91x_rsp_wireless_info_t wlan_info = { };
 	struct siwx91x_dev *sidev = dev->data;
+	uint8_t join_feature_bitmap = 0;
```
This variable will set set unconditionally by `sl_si91x_get_join_configuration()`, so it is not required to initialize it.

This is the result of `sl_si91x_get_join_configuration()`. So, unless you have a good reason, I would call it `join_config`.

Nitin-Pandey-01 replied:
Sure, Updated the code as suggested

Nitin-Pandey-01 commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -278,14 +278,22 @@ static int siwx91x_connect(const struct device *dev, struct wifi_connect_req_par
+					    SL_NET_WIFI_PSK, params->sae_password,
+					    params->sae_password_length);
+	} else if (params->psk_length) {
+		ret = sl_net_set_credential(SL_NET_DEFAULT_WIFI_CLIENT_CREDENTIAL_ID,
```
This case will get rejected in wifi_connect() function, it will not reach siwx91x_connect function.

Nitin-Pandey-01 commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -278,14 +278,22 @@ static int siwx91x_connect(const struct device *dev, struct wifi_connect_req_par
+					    SL_NET_WIFI_PSK, params->psk, params->psk_length);
+	}
+
+	if (ret != SL_STATUS_OK) {
```
Yes, It will work, but for WIFI_SECURITY_TYPE_NONE case.

- Shell will give a warning related to this.
- PSK for Open security mode will be ignored by firmware and connection will happen.

jerome-pouiller replied:
In case of `WIFI_SECURITY_TYPE_NONE`, the user won't provide any psk/sae_password. Is it still fine?

Nitin-Pandey-01 replied:
Got your point, Handled the case for OPEN security mode

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -153,6 +209,13 @@ static int siwx91x_ap_enable(const struct device *dev, struct wifi_connect_req_p
+	if (params->mfp != mfp_conf) {
+		LOG_WRN("Needed MFP %s but got MFP %s, hence setting to MFP %s",
+			wifi_mfp_txt(mfp_conf), wifi_mfp_txt(params->mfp), wifi_mfp_txt(mfp_conf));
+	}
```
I believe the case of the AP is so simple, you can handle it without the help of `siwx91x_configure_mfp_option()`:
```
if (params->mfp != WIFI_MFP_DISABLE) {
    LOG_WRN(...);
}
```
Then, you can change `siwx91x_configure_mfp_option()` in `siwx91x_configure_sta_mfp_option()` and save one level of indentation.

Nitin-Pandey-01 replied:
Sure, Updated the code as suggested

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -70,6 +70,57 @@ static int siwx91x_map_ap_security(enum wifi_security_type security)
+			sl_si91x_get_join_configuration(SL_WIFI_CLIENT_INTERFACE, &join_config);
+			sl_si91x_set_join_configuration(
+				SL_WIFI_CLIENT_INTERFACE,
+				(join_config | SL_SI91X_JOIN_FEAT_MFP_CAPABLE_REQUIRED));
```
The arguments should be aligned with the opened parenthesis (and prefer multiple statements rather than one statement on multiple lines):
```
			sl_si91x_get_join_configuration(SL_WIFI_CLIENT_INTERFACE, &join_config);
			join_config |= SL_SI91X_JOIN_FEAT_MFP_CAPABLE_REQUIRED;
			sl_si91x_set_join_configuration(SL_WIFI_CLIENT_INTERFACE, join_config);
```

Nitin-Pandey-01 replied:
Sure, Updated the code as suggested

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -290,6 +347,15 @@ static int siwx91x_connect(const struct device *dev, struct wifi_connect_req_par
+	mfp_conf = siwx91x_configure_sta_mfp_option(
+		wifi_config.security, (params->security == WIFI_SECURITY_TYPE_PSK_SHA256)
+					      ? WIFI_MFP_REQUIRED
+					      : params->mfp);
```
Let's don't abuse of the ternary operator:
```
	if (params->security == WIFI_SECURITY_TYPE_PSK_SHA256) {
		mfp_conf = siwx91x_set_sta_mfp_option(wifi_config.security, params->mfp);
        } else {
		mfp_conf = siwx91x_set_sta_mfp_option(wifi_config.security, WIFI_MFP_REQUIRED);
	}
```
(`set` is shorter than `configure` and is still meaningful)

Nitin-Pandey-01 replied:
Sure, Updated the code as suggested

jerome-pouiller reviewed (commented):
Two small cosmetic comments.

nipandey-01 committed:
driver: wifi: siwx91x: Add check for SAE password

- Modified conditions in WIFI_CONNECT()
  function to reject SAE password and
  PSK based on length parameter

Signed-off-by: Nitin Pandey <nitin.pandey@silabs.com>

nipandey-01 committed:
driver: wifi: siwx91x: Handle MFP configuration

- Handled MFP configuration for STA and AP modes
- Updated the MFP configuration based on security modes

Signed-off-by: Nitin Pandey <nitin.pandey@silabs.com>

Nitin-Pandey-01 commented:
@jhedberg Can you please review this PR.

jerome-pouiller commented:
@jukkar, I believe this PR can be merged.
